misc/metrics/src/swarm.rs

```diff
@@ -156,8 +157,10 @@ impl<TBvEv, THandleErr> super::Recorder<libp2p_swarm::SwarmEvent<TBvEv, THandleE
                     })
                     .inc();
             }
-            libp2p_swarm::SwarmEvent::IncomingConnection { .. } => {
-                self.connections_incoming.inc();
+            libp2p_swarm::SwarmEvent::IncomingConnection { send_back_addr, .. } => {
```

Comment:
How about doing the same for all other `SwarmEvent`s? E.g. `ConnectionEstablished`?

Reply:
I added ConnectionEstablished, but do you really want to do _all_ of them, even things like `connected_to_banned_peer` ?

````
$for u in http://127.0.0.1:37857/metrics http://127.0.0.1:34293/metrics ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 6
libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
````

Reply:
Added what I easily can, but these are triggered by events that don't seem to carry the an obvious multiaddr to key off of in the event struct itself:

dial_attempt
outgoing_connection_error
listener_error

These copy-pasted outputs are from the example running in my branch. Using exclusively tcp multiaddrs, hence the grep to find metrics keyed off protocol.
````
$for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
````

Reply:
> Added what I easily can, but these are triggered by events that don't seem to carry the an obvious multiaddr to key off of in the event struct itself:
> 
> - dial_attempt
> - outgoing_connection_error
> - listener_error

Yeah that is fine.

> These copy-pasted outputs are from the example running in my branch. Using exclusively tcp multiaddrs, hence the grep to find metrics keyed off protocol.
> 
> ```
> $for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done
> libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_incoming_total{address_stack="/ip4/tcp"} 1
> libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_established_total{role="Listener",proto_stack="/ip4/tcp"} 1
> libp2p_identify_listen_addresses_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_new_listen_addr_total{address_stack="/ip4/tcp"} 5
> libp2p_swarm_connections_established_total{role="Dialer",proto_stack="/ip4/tcp"} 1
> ```

We don't seem to be consistent in the naming. I'd actually prefer the "protocol"-inspired naming. How about we just call this label `protocols`:

```bash
$ for u in http://127.0.0.1:{37961,33459}/metrics  ; do curl -s "${u}" | grep tcp ; done

libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{protocols="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",protocols="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",protocols="/ip4/tcp"} 1
```

Reply:
I agree, @thomaseizinger . Now have:

```bash
$for u in http://127.0.0.1:{39507,42089}/metrics  ; do curl -s "${u}" | grep tcp ; done
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_incoming_total{protocols="/ip4/tcp"} 1
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Listener",protocols="/ip4/tcp"} 1
libp2p_identify_listen_addresses_total{protocols="/ip4/tcp"} 5
libp2p_swarm_new_listen_addr_total{protocols="/ip4/tcp"} 5
libp2p_swarm_connections_established_total{role="Dialer",protocols="/ip4/tcp"} 1
```